File: np_platform_image_format_jxl/android/src/main/kotlin/com/nkming/nc_photos/np_platform_image_format_jxl/NpPlatformImageFormatJxlPlugin.kt
package com.nkming.nc_photos.np_platform_image_format_jxl

import android.graphics.Bitmap
import com.awxkee.jxlcoder.JxlCoder
import com.awxkee.jxlcoder.ScaleMode
import com.nkming.nc_photos.np_android_core.Rgba8Image
import com.nkming.nc_photos.np_android_core.logE
import com.nkming.nc_photos.np_android_core.use
import io.flutter.embedding.engine.plugins.FlutterPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withPermit
import java.io.File

private class PigeonApiImpl : MyHostApi, CoroutineScope by MainScope() {
	companion object {
		private const val TAG = "NpPlatformImageFormatJxlPlugin"

		// Limit number of jobs to avoid overloading the phone
		private val lock = Semaphore(3)
	}

	override fun load(
		filepath: String, w: Long?, h: Long?, callback: (Result<Image>) -> Unit
	) {
		launch(Dispatchers.IO) {
			try {
				val result =
					decodeBytesToImage(File(filepath).readBytes(), w, h)
				callback(Result.success(result))
			} catch (e: Throwable) {
				callback(Result.failure(e))
			}
		}
	}

	override fun loadBytes(
		bytes: ByteArray, w: Long?, h: Long?, callback: (Result<Image>) -> Unit
	) {
		launch(Dispatchers.IO) {
			try {
				val result = decodeBytesToImage(bytes, w, h)
				callback(Result.success(result))
			} catch (e: Throwable) {
				callback(Result.failure(e))
			}
		}
	}

	override fun loadMetadata(
		filepath: String, callback: (Result<Metadata?>) -> Unit
	) {
		launch(Dispatchers.IO) {
			val size = try {
				JxlCoder.getSize(File(filepath).readBytes())
			} catch (e: Throwable) {
				logE(TAG, "Failed to decode file: $filepath", e)
				callback(
					Result.failure(
						FlutterError(
							"decode-error", "Failed to decode file", e.message
						)
					)
				)
				return@launch
			}
			if (size == null) {
				callback(Result.success(null))
			} else {
				callback(
					Result.success(
						Metadata(size.width.toLong(), size.height.toLong())
					)
				)
			}
		}
	}

	override fun save(
		img: Image, filepath: String, callback: (Result<Boolean>) -> Unit
	) {
		// TODO
		callback(Result.success(false))
	}

	override fun convertJpeg(
		filepath: String, w: Long?, h: Long?,
		callback: (Result<Unit>) -> Unit
	) {
		launch(Dispatchers.IO) {
			try {
				val bitmap = decodeBytes(File(filepath).readBytes(), w, h)
				bitmap.use {
					File(filepath).outputStream().use { oStream ->
						bitmap.compress(
							Bitmap.CompressFormat.JPEG, 85, oStream
						)
					}
				}
				callback(Result.success(Unit))
			} catch (e: Throwable) {
				callback(Result.failure(e))
			}
		}
	}

	private suspend fun decodeBytesToImage(
		bytes: ByteArray, w: Long?, h: Long?
	): Image {
		try {
			val bitmap = decodeBytes(bytes, w, h)
			val rgba8 = bitmap.use { Rgba8Image.fromBitmap(it) }
			return Image(
				pixel = rgba8.pixel,
				width = rgba8.width.toLong(),
				height = rgba8.height.toLong()
			)
		} catch (e: Throwable) {
			logE(TAG, "Failed to convert decoded data", e)
			throw FlutterError(
				"convert-error", "Failed to convert decoded data",
				e.message
			)
		}
	}

	private suspend fun decodeBytes(
		bytes: ByteArray, w: Long?, h: Long?
	): Bitmap {
		lock.withPermit {
			val bitmap = try {
				if (w != null && h != null) {
					JxlCoder.decodeSampled(
						bytes,
						w.toInt(),
						h.toInt(),
						scaleMode = ScaleMode.FIT
					)
				} else {
					JxlCoder.decode(bytes)
				}
			} catch (e: Throwable) {
				logE(TAG, "Failed to decode bytes", e)
				throw FlutterError(
					"decode-error", "Failed to decode bytes", e.message
				)
			}
			return bitmap
		}
	}
}

class NpPlatformImageFormatJxlPlugin : FlutterPlugin {
	override fun onAttachedToEngine(
		flutterPluginBinding: FlutterPlugin.FlutterPluginBinding
	) {
		val api = PigeonApiImpl()
		MyHostApi.setUp(flutterPluginBinding.binaryMessenger, api)
	}

	override fun onDetachedFromEngine(
		binding: FlutterPlugin.FlutterPluginBinding
	) {
	}
}

File: np_platform_image_format_jxl/ios/Classes/NpPlatformImageFormatJxlPlugin.swift
import Flutter
import JxlCoder
import Logging
import NpIosCore
import UIKit

public class NpPlatformImageFormatJxlPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let messenger = registrar.messenger()
    let api = NpPlatformImageFormatJxlPlugin()
    MyHostApiSetup.setUp(binaryMessenger: messenger, api: api)
  }
}

extension NpPlatformImageFormatJxlPlugin: MyHostApi {
  func load(
    filepath: String, w: Int64?, h: Int64?, completion: @escaping (Result<Image, Error>) -> Void
  ) {
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        let uiImage: UIImage
        if let w = w, let h = h {
          let size = try JXLCoder.getSize(url: URL.init(filePath: filepath))
          let dstSize = Self.fitSize(size: size, bbox: CGSize(width: Int(w), height: Int(h)))
          uiImage = try JXLCoder.decode(url: URL.init(filePath: filepath), rescale: dstSize)
        } else {
          uiImage = try JXLCoder.decode(url: URL.init(filePath: filepath))
        }
        let rgba8 = try Rgba8Image(uiImage: uiImage)
        let result = Image(
          pixel: rgba8.pixel.toFlutterTypedData(), width: Int64(rgba8.width),
          height: Int64(rgba8.height))
        completion(.success(result))
      } catch let e {
        Self.log.error("Failed to decode file: \(filepath) (\(e.localizedDescription))")
        completion(
          .failure(
            PigeonError(
              code: "decode-error", message: "Failed to decode file",
              details: e.localizedDescription)))
      }
    }
  }

  func loadBytes(
    bytes: FlutterStandardTypedData, w: Int64?, h: Int64?,
    completion: @escaping (Result<Image, Error>) -> Void
  ) {
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        let uiImage: UIImage
        if let w = w, let h = h {
          let size = try JXLCoder.getSize(data: bytes.data)
          let dstSize = Self.fitSize(size: size, bbox: CGSize(width: Int(w), height: Int(h)))
          uiImage = try JXLCoder.decode(data: bytes.data, rescale: dstSize)
        } else {
          uiImage = try JXLCoder.decode(data: bytes.data)
        }
        let rgba8 = try Rgba8Image(uiImage: uiImage)
        let result = Image(
          pixel: rgba8.pixel.toFlutterTypedData(), width: Int64(rgba8.width),
          height: Int64(rgba8.height))
        completion(.success(result))
      } catch let e {
        Self.log.error("Failed to decode data (\(e.localizedDescription))")
        completion(
          .failure(
            PigeonError(
              code: "decode-error", message: "Failed to decode data",
              details: e.localizedDescription)))
      }
    }
  }

  func loadMetadata(filepath: String, completion: @escaping (Result<Metadata?, Error>) -> Void) {
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        let size = try JXLCoder.getSize(url: URL.init(filePath: filepath))
        let result = Metadata(w: Int64(size.width), h: Int64(size.height))
        completion(.success(result))
      } catch let e {
        Self.log.error("Failed to decode file: \(filepath) (\(e.localizedDescription))")
        completion(
          .failure(
            PigeonError(
              code: "decode-error", message: "Failed to decode file",
              details: e.localizedDescription)))
      }
    }
  }

  func save(img: Image, filepath: String, completion: @escaping (Result<Bool, Error>) -> Void) {
    // TODO: Implement JXL image saving
    completion(
      .failure(
        PigeonError(code: "NOT_IMPLEMENTED", message: "save method not implemented", details: nil)))
  }

  func convertJpeg(
    filepath: String, w: Int64?, h: Int64?, completion: @escaping (Result<Void, Error>) -> Void
  ) {
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        let uiImage: UIImage
        if let w = w, let h = h {
          let size = try JXLCoder.getSize(url: URL.init(filePath: filepath))
          let dstSize = Self.fitSize(size: size, bbox: CGSize(width: Int(w), height: Int(h)))
          uiImage = try JXLCoder.decode(url: URL.init(filePath: filepath), rescale: dstSize)
        } else {
          uiImage = try JXLCoder.decode(url: URL.init(filePath: filepath))
        }
        guard let jpegData = uiImage.jpegData(compressionQuality: 0.85) else {
          Self.log.error("Failed to compress as JPEG: \(filepath)")
          completion(
            .failure(
              PigeonError(
                code: "compress-error", message: "Failed to compress as JPEG", details: nil)))
          return
        }
        do {
          try jpegData.write(to: URL(filePath: filepath))
          completion(.success(()))
        } catch let e {
          Self.log.error("Failed to write JPEG file: \(filepath) (\(e.localizedDescription))")
          completion(
            .failure(
              PigeonError(
                code: "compress-error", message: "Failed to write JPEG",
                details: e.localizedDescription)))
        }
      } catch let e {
        Self.log.error("Failed to decode file: \(filepath) (\(e.localizedDescription))")
        completion(
          .failure(
            PigeonError(
              code: "decode-error", message: "Failed to decode file",
              details: e.localizedDescription)))
      }
    }
  }

  private static func fitSize(size: CGSize, bbox: CGSize) -> CGSize {
    let aspectRatio = size.width / size.height
    let bboxAspectRatio = bbox.width / bbox.height

    if aspectRatio > bboxAspectRatio {
      // Fit by width
      return CGSize(width: bbox.width, height: bbox.width / aspectRatio)
    } else {
      // Fit by height
      return CGSize(width: bbox.height * aspectRatio, height: bbox.height)
    }
  }

  private static let log = Logger(label: "NpPlatformImageFormatJxlPlugin")
}
